lib/datadog/appsec/contrib/rack/gateway/response.rb

```diff
@@ -19,9 +19,58 @@ def initialize(body, status, headers, scope:)
               @scope = scope
             end
 
+            def parsed_body
+              return unless Datadog.configuration.appsec.parse_response_body
+
+              unless body.instance_of?(Array) || body.instance_of?(::Rack::BodyProxy)
+                if result.timeout
+                  Datadog.logger.debug do
+                    "Unable to parse response body because of unsupported body type: #{body.class}"
+                  end
+                end
+                return
+              end
+              return unless supported_response_type
+
+              body_dup = body.dup # avoid interating over the body. This is just in case code.
```

Comment:
For `Array` this is not needed, so is that for `Rack::BodyProxy`?

Reply:
Yes. 

Since `Rack::BodyProxy` acts as a proxy. If we call `each` directly, we might iterate over the body and consume it. 

https://github.com/rack/rack/blob/main/lib/rack/body_proxy.rb#L45-L58

If we call to `to_ary` we also close the body, and we do not want that, since we might not be the last middleware. That is why I thought using `dup` would save us here

Reply:
But then you're duping the `Rack::BodyProxy` instance, which would still point to the same underlying wrapped object, which we don't know the type of, and may not be traversable twice.